cmd/agent/app/reporter/client_metrics_test.go

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/uber/jaeger-lib/metrics/metricstest"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest/observer"
+
+	"github.com/jaegertracing/jaeger/cmd/agent/app/testutils"
+	"github.com/jaegertracing/jaeger/thrift-gen/jaeger"
+	"github.com/jaegertracing/jaeger/thrift-gen/zipkincore"
+)
+
+type clientMetricsTest struct {
+	mr   *testutils.InMemoryReporter
+	r    *ClientMetricsReporter
+	logs *observer.ObservedLogs
+	mb   *metricstest.Factory
+}
+
+func (tr *clientMetricsTest) assertLog(t *testing.T, msg, clientUUID string) {
+	logs := tr.logs.FilterMessageSnippet(msg)
+	if clientUUID == "" {
+		assert.Equal(t, 0, logs.Len(), "not expecting log '%s", msg)
+	} else {
+		if assert.Equal(t, 1, logs.Len(), "expecting one log '%s'", msg) {
+			field := logs.All()[0].ContextMap()["client-uuid"]
+			assert.Equal(t, clientUUID, field, "client-uuid should be logged")
+		}
+	}
+}
+
+func testClientMetrics(fn func(tr *clientMetricsTest)) {
+	testClientMetricsWithParams(ClientMetricsReporterParams{}, fn)
+}
+
+func testClientMetricsWithParams(params ClientMetricsReporterParams, fn func(tr *clientMetricsTest)) {
+	r1 := testutils.NewInMemoryReporter()
+	zapCore, logs := observer.New(zap.DebugLevel)
+	mb := metricstest.NewFactory(time.Hour)
+
+	params.Reporter = r1
+	params.Logger = zap.New(zapCore)
+	params.MetricsFactory = mb
+
+	r := WrapWithClientMetrics(params)
+	defer r.Close()
+
+	tr := &clientMetricsTest{
+		mr:   r1,
+		r:    r,
+		logs: logs,
+		mb:   mb,
+	}
+	fn(tr)
+}
+
+func TestClientMetricsReporter_Zipkin(t *testing.T) {
+	testClientMetrics(func(tr *clientMetricsTest) {
+		assert.NoError(t, tr.r.EmitZipkinBatch([]*zipkincore.Span{{}}))
+		assert.Len(t, tr.mr.ZipkinSpans(), 1)
+	})
+}
+
+func TestClientMetricsReporter_Jaeger(t *testing.T) {
+	testClientMetrics(func(tr *clientMetricsTest) {
+		blank := ""
+		clientUUID := "foobar"
```

Comment:
Are the known clients supposed to send values compliant with the RFC-4122? If so, should the agent require the format to be compliant with that? I can't think of a good reason for a client to send a non-compliant value, other than configuration error...

Reply:
I don't think they are sending real UUIDs today (probably just random bits). We just need them to be reasonably unique. As far as this code is concerned, it's just some opaque string.

Reply:
Alright. If we knew that all current clients were sending real UUIDs, we could assume that anything non-standard would be a misconfigured client.